libazureinit/src/provision/ssh.rs

```diff
@@ -43,123 +45,225 @@ pub(crate) fn provision_ssh(
     Ok(())
 }
 
-pub(crate) fn get_authorized_keys_path_from_sshd() -> io::Result<Option<PathBuf>>
-{
-    let sshd_output = Command::new("sshd").arg("-T").output()?;
-    if !sshd_output.status.success() {
-        return Err(io::Error::new(
-            io::ErrorKind::Other,
-            format!("sshd -T failed with status: {}", sshd_output.status),
-        ));
+fn get_sshd_g_output() -> io::Result<Output> {
+    Command::new("sshd").arg("-G").output()
+}
+
+fn get_authorized_keys_path_from_sshd(
+    sshd_config_command_runner: impl Fn() -> io::Result<Output>,
+) -> Option<String> {
+    let output: Output = match run_sshd_command(sshd_config_command_runner) {
+        Some(output) => output,
+        None => {
+            info!("Unable to determine authorizedkeysfile from sshd -G");
+            return None;
+        }
+    };
+
+    match extract_authorized_keys_file_path(&output.stdout) {
+        Some(path) => Some(path),
+        None => {
+            info!("No authorizedkeysfile in sshd configuration");
+            None
+        }
     }
+}
 
-    let stdout = sshd_output.stdout;
-    let sshd_output = String::from_utf8_lossy(&stdout);
-    for line in sshd_output.lines() {
+fn run_sshd_command(
+    sshd_config_command_runner: impl Fn() -> io::Result<Output>,
+) -> Option<Output> {
+    match sshd_config_command_runner() {
+        Ok(output) if output.status.success() => {
+            info!(
+                "sshd -G succeeded with stdout length={:?}",
+                output.stdout.len()
+            );
+            Some(output)
+        }
+        Ok(output) => {
+            error!(
+                "sshd -G failed with code={} stdout={:?} stderr={:?}",
+                output.status,
+                String::from_utf8_lossy(&output.stdout),
+                String::from_utf8_lossy(&output.stderr)
+            );
+            None
+        }
+        Err(e) => {
+            error!("Failed to execute sshd -G due to {}", e);
+            None
+        }
+    }
+}
+
+fn extract_authorized_keys_file_path(stdout: &[u8]) -> Option<String> {
+    let output = String::from_utf8_lossy(stdout);
+    for line in output.lines() {
         if line.starts_with("authorizedkeysfile") {
             let keypath: Vec<&str> = line.split_whitespace().collect();
             if keypath.len() > 1 {
-                info!("Found authorized_keys path: {}", keypath[1]);
-                return Ok(Some(PathBuf::from(keypath[1])));
+                info!("Found authorizedkeysfile path: {}", keypath[1]);
+                return Some(String::from(keypath[1]));
```

Comment:
I know you're just refactoring this, but there's a way to write this without doing any sort of length checking and explicit indexing into the vector:

```Rust
    for line in output.lines() {
        if line.starts_with("authorizedkeysfile") {
            let keypath = line.split_whitespace().skip(1).next().map(|s| {
                info!("Found authorizedkeysfile path: {}", s);
                s.to_string()
            });
            if keypath.is_some() {
                return keypath;
            }
        }
    }
```

Rather than collecting the iterator, we'll skip the first element (the setting name) and then operate on the next element if it's there.

Reply:
done, thanks!